Reference/Common-Pitfalls/index-v9.md

```diff
@@ -0,0 +1,426 @@
+
+The primary reasons your data will become out of sync are:
+
+* Old version of Umbraco
+* Rebuilding indexes and restarting your app domain at the same time (try to avoid this scenario!)
+
+It is not recommended to rebuild your indexes unless you absolutely need to and if you need to do this often then it is
+advised to determine why and to try to resolve the underlying problem.
+
+## Performing lookups and logic in Examine events
+
+There's a couple well known Examine events: `GatheringNodeData` and `DocumentWriting`. Both of these events
+allow the developer to modify the data that is going into the Lucene index but many times we see developers Performing
+Service lookups in these methods. For example, using `IContentService.GetById(e.NodeId)`
+inside of these events could cause an `N + 1` problem. This is because these events are executed for every single document
+being indexed and if you are rebuilding an index, this will mean this logic will fire for every single document and media item
+going into each index ... That could mean a tremendous number of lookups and performance drain.
+
+Similarly, if you are executing other logic in these events that perform poorly, then anytime you save or publish content or media
+it will slow that process down. And if you rebuild an index then any slow code running in these events will cause the indexing
+to go ultra slow.
+
+## RenderTemplateAsync
+
+There is an API in Umbraco that should never be used unless you really know what you are doing. This API method
+is called `RenderTemplateAsync`. It allows you to be able to render a particular content item's template and get a `IHtmlEncodedString`
+in response. In some cases, this may be useful. Perhaps you want to send an email based on a content item and its template, but
+you must be very careful not to use this for purposes it is not meant to be used for.
+
+Generally speaking this method should not be used for the normal rendering of content. If abused this could cause severe
+performance problems. For normal content rendering of module type data from another content item, you should use Partial Views instead.
+
+## Don't put logic inside your constructors
+
+Constructors should generally not perform any logic, they should set some parameter values, perform some null checks and perhaps validate some data but in most cases, they should not perform any logic.
+
+There's a few reasons why this can become a huge performance problem:
+
+* The consumer of an API doesn't expect that by creating an object that they should be worried about performance
+* Creating an object can inadvertently happen a vast number of times, especially when using LINQ
+
+Here's an example of how this can go wrong very quickly:
+Your tree structure is something like this:
+
+    - Root
+    -- Home
+    --- Recipes (node id = 3251, list view with 5000 nodes)
+    --- About Us
+    --- Contact Us
+
+You have a custom model that looks like:
```

Comment:
```suggestion
## Don't put logic inside your constructors

Constructors should generally not perform any logic, they should set some parameter values, perform some null checks and perhaps validate some data but in most cases, they should not perform any logic.

There are a few reasons why this can become a huge performance problem:

* The consumer of an API doesn't expect that by creating an object that they should be worried about performance
* Creating an object can inadvertently happen a vast number of times, especially when using LINQ

Here's an example of how this can go wrong very quickly:
Your tree structure is something like this:

    - Root
    -- Home
    --- Recipes (node id = 3251, list view with 5000 nodes)
    --- About Us
    --- Contact Us

You have a custom model that looks like this:
```